In [13]:

import pandas as pd
# Lendo os dados
geo_data_0 = pd.read_csv('geo_data_0.csv')
geo_data_1 = pd.read_csv('geo_data_1.csv')
geo_data_2 = pd.read_csv('geo_data_2.csv')
print(geo_data_0.head())
print(geo_data_1.head())
print(geo_data_2.head())


      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746


In [14]:
print(geo_data_0.info())
print(geo_data_1.info())
print(geo_data_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column  

In [15]:
print(geo_data_0.describe())
print(geo_data_1.describe())
print(geo_data_2.describe())

                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347
                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%       

In [16]:
print(geo_data_0.isnull().sum())
print(geo_data_1.isnull().sum())
print(geo_data_2.isnull().sum())

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64
id         0
f0         0
f1         0
f2         0
product    0
dtype: int64
id         0
f0         0
f1         0
f2         0
product    0
dtype: int64


In [17]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np


In [18]:
def load_data(file_path):
    """Carrega os dados de um arquivo CSV."""
    return pd.read_csv(file_path)

def train_and_evaluate_model(data):
    """Treina e avalia o modelo de regressão linear."""
    # Dividir os dados em características e alvo
    X = data[['f0', 'f1', 'f2']]
    y = data['product']
    
    # Dividir os dados em conjunto de treinamento e validação (75:25)
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.25, random_state=0)
    
    # Treinar o modelo
    model = LinearRegression()
    model.fit(X_train, y_train)
    
    # Fazer predições
    predictions = model.predict(X_val)
    
    # Calcular o volume médio previsto de reservas
    average_predicted = np.mean(predictions)
    
    # Calcular o REQM (Root Mean Squared Error)
    reqm = np.sqrt(mean_squared_error(y_val, predictions))
    
    # Salvar predições e respostas corretas em um DataFrame
    results = pd.DataFrame({'Actual': y_val, 'Predicted': predictions})
    
    return average_predicted, reqm, results

def analyze_results(results):
    """Analisa os resultados e imprime as informações relevantes."""
    print("Resultados da Análise:")
    print(results.head())
    print("\nVolume médio previsto de reservas:", results['Predicted'].mean())
    print("REQM do modelo:", np.sqrt(mean_squared_error(results['Actual'], results['Predicted'])))

In [19]:
# Lista de arquivos
files = ['geo_data_0.csv', 'geo_data_1.csv', 'geo_data_2.csv']

for file in files:
    print(f"\nAnalisando {file}:")
    data = load_data(file)
    average_predicted, reqm, results = train_and_evaluate_model(data)
    analyze_results(results)


Analisando geo_data_0.csv:
Resultados da Análise:
           Actual   Predicted
3582    49.723143   95.303891
60498   10.583795   95.098407
53227  139.110613  106.702273
21333   85.109875   96.740059
3885    39.726827   97.703381

Volume médio previsto de reservas: 92.27144852242301
REQM do modelo: 37.48100896950594

Analisando geo_data_1.csv:
Resultados da Análise:
           Actual   Predicted
3582   134.766305  134.451381
60498  137.945408  137.031555
53227    3.179103    2.835624
21333  110.992147  111.274297
3885   137.945408  138.109072

Volume médio previsto de reservas: 69.15162398290752
REQM do modelo: 0.8872573052219331

Analisando geo_data_2.csv:
Resultados da Análise:
           Actual   Predicted
3582   129.102262  103.879759
60498   61.285744   59.948159
53227   65.026246  108.629782
21333   62.334048   99.219377
3885   150.002134   90.361944

Volume médio previsto de reservas: 94.70753129105672
REQM do modelo: 40.31290686044374


geo_0= prevê um volume médio de reservas de aproximadamente 92.27 mil barris e No o REQM de 37.48, que indica que há uma quantidade considerável de erro nas previsões; 
geo_1= o volume médio previsto é mais baixo, em torno de 69.15 mil barris, mas o REQM é muito menor (0.89), indica que está mais próximo dos valores reais;
geo_2= erros consideráveis também, parecido com o 0

In [20]:

# Definindo os valores
investment_total = 100_000_000  # 100 milhões de dólares
number_of_wells = 200  # número de poços
required_revenue_per_well = 500_000  # receita necessária por poço em dólares

# Cálculo da quantidade de unidades necessárias por poço
required_units_per_well = required_revenue_per_well / 4.5  # considerando que cada barril gera 4.5 dólares


In [21]:

# Valores médios previstos para cada região
average_volumes = {
    'geo_data_0': 92.27,  # em milhares de barris
    'geo_data_1': 69.15,
    'geo_data_2': 94.71
}

# Comparação
for region, average_volume in average_volumes.items():
    average_volume_units = average_volume * 1000  # Convertendo para unidades
    if average_volume_units >= required_units_per_well:
        conclusion = "suficiente"
    else:
        conclusion = "insuficiente"
    
    print(f"Região: {region}, Volume médio previsto: {average_volume_units:.2f} unidades, "
          f"Necessário: {required_units_per_well:.2f} unidades. Conclusão: {conclusion}.")


Região: geo_data_0, Volume médio previsto: 92270.00 unidades, Necessário: 111111.11 unidades. Conclusão: insuficiente.
Região: geo_data_1, Volume médio previsto: 69150.00 unidades, Necessário: 111111.11 unidades. Conclusão: insuficiente.
Região: geo_data_2, Volume médio previsto: 94710.00 unidades, Necessário: 111111.11 unidades. Conclusão: insuficiente.


geo_0= O volume médio previsto não atinge o mínimo necessário para cobrir os custos de operação
geo_1= baixa viabilidade econômica também, volume de reservas abaixo
geo_2= mesmo com um volume médio razoavelmente alto, esta região ainda não atinge o volume necessário para evitar prejuízos, não é atraente para o investimento. 
Concluindo= Todas as regiões analisadas mostram volumes médios previstos insuficientes para cobrir

In [22]:
def calculate_profit(data, price_per_barrel=4.5, num_wells=200):
    """
    Calcula o lucro potencial dos poços de petróleo com base nas predições do modelo.

    :param data: DataFrame contendo as predições e os volumes de reservas reais.
    :param price_per_barrel: Preço por barril de petróleo.
    :param num_wells: Número de poços a serem considerados para o cálculo.
    :return: lucro potencial e as predições dos poços selecionados.
    """
    # Selecionar os 200 poços com os volumes previstos mais altos
    top_wells = data.nlargest(num_wells, 'Predicted')
    
    # Calcular o lucro potencial
    total_volume = top_wells['Predicted'].sum() * 1000  # Convertendo para unidades (barris)
    total_revenue = total_volume * price_per_barrel
    total_cost = 100_000_000  # Custo total do investimento
    profit = total_revenue - total_cost
    
    return profit, top_wells

In [23]:
# Resultado para cada região
results = {}

# Lista de arquivos e médias de volume
files = ['geo_data_0.csv', 'geo_data_1.csv', 'geo_data_2.csv']
average_volumes = {
    'geo_data_0': 92.27,
    'geo_data_1': 69.15,
    'geo_data_2': 94.71
}

# Analisando cada região
for file in files:
    # Carregar os dados
    data = load_data(file)
    
    # Treinar o modelo e obter predições
    _, _, results_df = train_and_evaluate_model(data)
    
    # Calcular lucro potencial
    profit, top_wells = calculate_profit(results_df)
    
    # Armazenar resultados
    results[file] = {
        'profit': profit,
        'top_wells': top_wells
    }

In [24]:

# Analisando os resultados
for file, result in results.items():
    print(f"\nResultados para {file}:")
    print(f"Lucro potencial: ${result['profit']:.2f}")
    print("Top 200 poços selecionados:")
    print(result['top_wells'][['Actual', 'Predicted']].head())

# Conclusão e recomendação
best_region = max(results.items(), key=lambda x: x[1]['profit'])
print(f"\nRecomendação: Desenvolver poços na região {best_region[0]} com lucro potencial de ${best_region[1]['profit']:.2f}.")



Resultados para geo_data_0.csv:
Lucro potencial: $39312763.79
Top 200 poços selecionados:
           Actual   Predicted
93073  162.810993  180.180276
46784  153.639837  176.258518
93716  130.985681  172.541307
57591  124.949574  170.199538
74946  168.479729  168.766978

Resultados para geo_data_1.csv:
Lucro potencial: $24883827.79
Top 200 poços selecionados:
           Actual   Predicted
80439  137.945408  139.983513
84991  137.945408  139.587723
60464  137.945408  139.568807
24274  137.945408  139.553702
81453  137.945408  139.524121

Resultados para geo_data_2.csv:
Lucro potencial: $33918196.40
Top 200 poços selecionados:
           Actual   Predicted
98619  175.103291  166.272188
70337   84.567996  166.175939
91268  101.014932  165.907515
54085   92.947333  165.422313
17415  142.806484  162.637379

Recomendação: Desenvolver poços na região geo_data_0.csv com lucro potencial de $39312763.79.


In [25]:

import numpy as np

def bootstrap_profit_distribution(top_wells, price_per_barrel=4.5, num_samples=1000):
    """
    Realiza bootstrapping para calcular a distribuição de lucros.

    :param top_wells: DataFrame com os 200 poços selecionados.
    :param price_per_barrel: Preço por barril de petróleo.
    :param num_samples: Número de amostras para o bootstrapping.
    :return: lucro médio, intervalo de confiança de 95% e probabilidade de prejuízo.
    """
    profits = []
    
    # Cálculo do lucro real para os 200 poços
    total_volume = top_wells['Predicted'].sum() * 1000  # Convertendo para unidades (barris)
    total_revenue = total_volume * price_per_barrel
    total_cost = 100_000_000  # Custo total do investimento
    initial_profit = total_revenue - total_cost
    
    # Realizando bootstrapping
    for _ in range(num_samples):
        sample = np.random.choice(top_wells['Predicted'], size=len(top_wells), replace=True)
        total_volume_sample = sample.sum() * 1000
        total_revenue_sample = total_volume_sample * price_per_barrel
        profit_sample = total_revenue_sample - total_cost
        profits.append(profit_sample)
    
    # Cálculo do lucro médio e intervalo de confiança
    mean_profit = np.mean(profits)
    confidence_interval = np.percentile(profits, [2.5, 97.5])
    
    # Cálculo da probabilidade de prejuízo
    risk_of_loss = np.mean(np.array(profits) < 0) * 100  # Expressando como porcentagem
    
    return mean_profit, confidence_interval, risk_of_loss


In [26]:
# Resultado para cada região com bootstrapping
bootstrap_results = {}

for file, result in results.items():
    mean_profit, confidence_interval, risk_of_loss = bootstrap_profit_distribution(result['top_wells'])
    bootstrap_results[file] = {
        'mean_profit': mean_profit,
        'confidence_interval': confidence_interval,
        'risk_of_loss': risk_of_loss
    }

In [27]:

# Analisando os resultados de bootstrapping
for file, stats in bootstrap_results.items():
    print(f"\nResultados de Bootstrapping para {file}:")
    print(f"Lucro médio: ${stats['mean_profit']:.2f}")
    print(f"Intervalo de Confiança de 95%: (${stats['confidence_interval'][0]:.2f}, ${stats['confidence_interval'][1]:.2f})")
    print(f"Risco de Prejuízo: {stats['risk_of_loss']:.2f}%")

# Conclusão e recomendação
best_region_bootstrap = max(bootstrap_results.items(), key=lambda x: x[1]['mean_profit'])
print(f"\nRecomendação: Desenvolver poços na região {best_region_bootstrap[0]} com lucro médio de ${best_region_bootstrap[1]['mean_profit']:.2f}.")



Resultados de Bootstrapping para geo_data_0.csv:
Lucro médio: $39326196.31
Intervalo de Confiança de 95%: ($38589180.69, $40035952.38)
Risco de Prejuízo: 0.00%

Resultados de Bootstrapping para geo_data_1.csv:
Lucro médio: $24883192.74
Intervalo de Confiança de 95%: ($24847624.22, $24919312.60)
Risco de Prejuízo: 0.00%

Resultados de Bootstrapping para geo_data_2.csv:
Lucro médio: $33933753.24
Intervalo de Confiança de 95%: ($33284544.89, $34606118.40)
Risco de Prejuízo: 0.00%

Recomendação: Desenvolver poços na região geo_data_0.csv com lucro médio de $39326196.31.


Recomendação de desenvolvimento de poços de petróleo na região geo_0
O lucro médio para esta região é o mais alto entre as três, com um intervalo de confiança estreito, indicando uma previsão confiável. A ausência de risco de prejuízo mostra que a exploração nessa área é altamente viável, garantindo retornos positivos. 

In [9]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

PRICE_PER_BARREL = 4.5
BUDGET = 100_000_000
N_STUDY = 500      # pontos estudados por rodada
N_SELECT = 200     # poços desenvolvidos
N_BOOT = 1000
RANDOM_STATE = 0   # para reprodutibilidade

def train_and_get_validation_frame(df, random_state=RANDOM_STATE):
    """Treina regressão linear e retorna um DataFrame de validação com product (real) e pred (previsto)."""
    X = df[['f0','f1','f2']]
    y = df['product']
    X_tr, X_val, y_tr, y_val = train_test_split(
        X, y, test_size=0.25, random_state=random_state
    )
    model = LinearRegression().fit(X_tr, y_tr)
    preds = model.predict(X_val)

    val_df = X_val.copy()
    val_df['product'] = y_val.values       # alvo real
    val_df['pred'] = preds                 # previsão do modelo
    return val_df

def profit_from_selected(selected_df, price_per_barrel=PRICE_PER_BARREL, budget=BUDGET):
    """
    Calcula lucro usando o product REAL dos poços selecionados.
    product está em milhares de barris -> multiplicar por 1000 e pelo preço por barril.
    """
    revenue = selected_df['product'].sum() * 1000 * price_per_barrel
    return revenue - budget

def bootstrap_region(val_df, 
                     n_boot=N_BOOT, 
                     n_study=N_STUDY, 
                     n_select=N_SELECT, 
                     seed=RANDOM_STATE):
    """
    Em cada rodada:
      1) amostra 500 pontos (com reposição) do conjunto de validação;
      2) escolhe top-200 por 'pred';
      3) calcula lucro com 'product' REAL desses 200.
    Retorna lucro médio, IC 95% e risco de prejuízo (%).
    """
    rng = np.random.default_rng(seed)
    profits = np.empty(n_boot)

    for i in range(n_boot):
        # amostra 500 com reposição
        sample_idx = rng.choice(val_df.index, size=n_study, replace=True)
        sample = val_df.loc[sample_idx]

        # escolhe top-200 por previsão
        top = sample.nlargest(n_select, 'pred')

        # lucro com product real
        profits[i] = profit_from_selected(top)

    mean_profit = profits.mean()
    ci_low, ci_high = np.percentile(profits, [2.5, 97.5])
    risk_of_loss = (profits < 0).mean() * 100.0
    return mean_profit, (ci_low, ci_high), risk_of_loss, profits

# --- Carrega dados (ajuste caminhos se necessário) ---
geo0 = pd.read_csv('geo_data_0.csv')
geo1 = pd.read_csv('geo_data_1.csv')
geo2 = pd.read_csv('geo_data_2.csv')

# --- Treina e prepara frames de validação com real e previsto ---
val0 = train_and_get_validation_frame(geo0)
val1 = train_and_get_validation_frame(geo1)
val2 = train_and_get_validation_frame(geo2)

# --- Bootstrapping por região ---
stats = {}
for name, val in [('geo_data_0', val0), ('geo_data_1', val1), ('geo_data_2', val2)]:
    mean_p, (lo, hi), risk, profits = bootstrap_region(val)
    stats[name] = {
        'mean_profit': mean_p,
        'ci95': (lo, hi),
        'risk_loss_pct': risk
    }

# --- Impressão dos resultados ---
for name, s in stats.items():
    print(f"\n{name}:")
    print(f"  Lucro médio: ${s['mean_profit']:,.2f}")
    print(f"  IC 95%: (${s['ci95'][0]:,.2f}, ${s['ci95'][1]:,.2f})")
    print(f"  Risco de prejuízo: {s['risk_loss_pct']:.2f}%")

# --- Decisão final: filtra regiões com risco < 2.5% e escolhe a de maior lucro médio ---
candidatas = {k:v for k,v in stats.items() if v['risk_loss_pct'] < 2.5}
if candidatas:
    melhor = max(candidatas.items(), key=lambda kv: kv[1]['mean_profit'])
    regiao_escolhida, s = melhor
    print(f"\nRecomendação FINAL: desenvolver na região {regiao_escolhida} "
          f"(lucro médio ${s['mean_profit']:,.2f}, IC95% {s['ci95']}, "
          f"risco {s['risk_loss_pct']:.2f}%).")
else:
    # se nenhuma atende o critério de risco, reporta a melhor no trade-off
    melhor = max(stats.items(), key=lambda kv: kv[1]['mean_profit'])
    regiao_escolhida, s = melhor
    print("\nNenhuma região ficou abaixo de 2,5% de risco. "
          f"Ainda assim, a melhor relação retorno/risco é {regiao_escolhida} "
          f"(lucro médio ${s['mean_profit']:,.2f}, IC95% {s['ci95']}, "
          f"risco {s['risk_loss_pct']:.2f}%).")



geo_data_0:
  Lucro médio: $4,388,001.22
  IC 95%: ($-577,879.76, $9,600,091.43)
  Risco de prejuízo: 4.30%

geo_data_1:
  Lucro médio: $4,749,577.11
  IC 95%: ($687,939.21, $8,567,600.67)
  Risco de prejuízo: 1.70%

geo_data_2:
  Lucro médio: $3,879,559.73
  IC 95%: ($-1,502,891.91, $9,034,400.97)
  Risco de prejuízo: 7.60%

Recomendação FINAL: desenvolver na região geo_data_1 (lucro médio $4,749,577.11, IC95% (687939.2058323596, 8567600.668736717), risco 1.70%).


Pelo lucro bruto (4.3), a região mais atraente parecia ser a geo_data_0, mas quando avaliamos risco e variabilidade (5), a única que atende ao critério (<2,5% de risco de prejuízo) é geo_data_1.
